document/core/exec/instructions.rst

```diff
@@ -737,243 +768,354 @@ Memory Instructions
+     S'; F; (t.\CONST~c_1)
    \end{array}
    \\ \qquad
      \begin{array}[t]{@{}r@{~}l@{}}
-     (\iff & \X{ea} + N/8 \leq n \\
+     (\iff & \meminst = S.\SMEMS[a] \\
+     \wedge & \X{ea} + N/8 \leq n \\
      \wedge & \X{ea} \mod N/8 = 0 \\
-     \wedge & b_1^\ast = \bytes_{\iN}(c_1) \\
-     \wedge & b^\ast = \bytes_{\iN}(\wrap_{|t|,N}(\atomicop_t(c_1, c_2)))) \\
+     \wedge & c_1 = \extendtu_{\X{st},t}(\bytes^{-1}_{\X{st}}(\meminst.\MIDATA[\X{ea} \slice N/8])) \\
+     \wedge & S' = S \with \SMEMS[a].\MIDATA[\X{ea} \slice N/8] = \bytes_{\X{st}}(\wrapt_{t,\X{st}}(\atop_t(c_1, c_2)))) \\
      \end{array}
    \\[1ex]
    \begin{array}{lcl@{\qquad}l}
-   F; (\I32.\CONST~k)~(t.\CONST~c)~(t.\ATOMICRMW({N}\K{\_u})^?.\atomicop~\memarg)
+   S; F; (\I32.\CONST~k)~(t.\CONST~c)~(t.\ATOMICRMW({N}\K{\_u})^?.\atop~\memarg)
+     &\stepto&
+     S; F; \TRAP
+   \end{array}
+   \\ \qquad
+     \begin{array}[t]{@{}r@{~}l@{}}
+     (\iff & \meminst = S.\SMEMS[a] \\
+     \wedge & \X{ea} + N/8 > n \vee \X{ea} \mod N/8 \neq 0) \\
+     \end{array}
+   \\
+   %
+   ~\\
+   \begin{array}{lcl@{\qquad}l}
+   S; F; (\I32.\CONST~i)~(t.\CONST~c_2)~(t.\ATOMICRMW({N}\K{\_u})^?.\atop~\memarg)
+     &\stepto^{(\ARD~a.\LLEN~n)~(\ARMW~a.\LDATA[\X{ea}]~b_{\F{r}}^\ast~b_{\F{w}}^\ast)}&
+     S; F; (t.\CONST~c_1)
+   \end{array}
+   \\ \qquad
+     \begin{array}[t]{@{}r@{~}l@{}}
+     (\iff & S.\SMEMS[a] ~\mbox{undefined} \\
+     \wedge & \X{ea} + N/8 \leq n \\
+     \wedge & \X{ea} \mod N/8 = 0 \\
+     \wedge & c_1 = \extendtu_{\X{st},t}(\bytes^{-1}_{\X{st}}(b_{\F{r}}^\ast)) \\
+     \wedge & b_{\F{r}}^\ast = \bytes_{\X{st}}(\wrapt_{t,\X{st}}(\atop_t(c_1, c_2)))) \\
+     \end{array}
+   \\[1ex]
+   \begin{array}{lcl@{\qquad}l}
+   S; F; (\I32.\CONST~k)~(t.\CONST~c)~(t.\ATOMICRMW({N}\K{\_u})^?.\atop~\memarg)
      &\stepto^{(\ARD~a.\LLEN~n)}&
-     F; \TRAP
+     S; F; \TRAP
    \end{array}
    \\ \qquad
      \begin{array}[t]{@{}r@{~}l@{}}
-     (\iff & \X{ea} + N/8 > n \vee \X{ea} \mod N/8 \neq 0) \\
+     (\iff & S.\SMEMS[a] ~\mbox{undefined} \\
+     \wedge & \X{ea} + N/8 > n \vee \X{ea} \mod N/8 \neq 0) \\
      \end{array}
-   \\[2ex]
-   (\where N = |t| \iff ~\mbox{not present}) \\
-   \\[2ex]
+   \\
+   %
+   ~\\
    \begin{array}[t]{@{}r@{~}l@{}}
-   (\where & N = |t| \iff ~\mbox{not present} \\
+   (\where & N = |t| \iff N~\mbox{not present} \\
+   \wedge & \X{st} = t \iff N~\mbox{not present}, ~\X{st} = \iN \otherwise \\
    \wedge & a = F.\AMODULE.\MIMEMS[0] \\
    \wedge & \X{ea} = i + \memarg.\OFFSET) \\
    \end{array}
    \end{array}
 
 
-.. _exec-atomic-rmw-cmpxchg:
-.. _exec-atomic-rmwn-cmpxchg:
+.. _exec-atomic.rmw.cmpxchg:
+.. _exec-atomic.rmwn.cmpxchg:
 
-:math:`t\K{.}\ATOMICRMW{.}\ATOMICCMPXCHG~\memarg` and :math:`t\K{.}\ATOMICRMW{N}\K{\_u.}\ATOMICCMPXCHG~\memarg`
-...............................................................................................................
+:math:`t\K{.}\ATOMICRMW({N}\K{\_u})^?\K{.}\ATCMPXCHG~\memarg`
```

Comment:
High-level:

cmpxchg should only perform an rmw action if the comparison succeeds. This means that at the operational level, the instruction non-deterministically either emits an atomic read action where the observed value is != the argument, or an rmw action where the read value is == the argument.

Reply:
Huh, the JS spec also models compareExchange as an rmw even if the comparison fails. I'll double check, but I'm pretty sure that's not correct.

Reply:
C++11 models it as just a load in the failure case, so I think this is also a JS spec bug.